timescale/promscale_extension
# Dump/restore fixes

# Test the dump/restore process

Added a test to ensure that the dump/restore process works.

To test this PR, run `make docker-image-14`, then run this command:

```
TS_DOCKER_IMAGE=local/dev_promscale_extension:head-ts2-pg14 cargo test --package e2e --test dump-restore-test dump_restore_test -- --exact
```

There is [a corresponding PR](https://github.com/timescale/promscale/pull/1311) on the promscale connector repo.

The rough outline of the test:

1. create a Postgres docker container with timescaledb
2. create the promscale extension
3. add metrics, exemplars, and traces to the database
4. snapshot the database
5. use pg_dump to create a logical SQL dump
6. destroy the docker container and create a new one
7. restore into the new database
8. snapshot the database
9. create more data to make sure things are still wired up correctly
10. compare the two snapshots. they should be equal

This uncovered several issues that were addressed:

The _ps_catalog.migration and _ps_catalog.promscale_instance_information tables should not be extension config tables. This means the structure of the tables will be dumped by pg_dump, but the data in them will not. The extension code will populate these tables appropriately. Because we are handling the _ps_catalog.migration table this way, you must use the same version of the extension on the restore side that was present on the dump side.

The extension inserts an initial row into the _prom_catalog.ids_epoch table on installation. We need to skip this if a restore is in progress so that the dumped value is what is inserted.

The public.prom_installation_info table was unused and caused problems for restores, so it was removed.

Both the _prom_catalog.remote_commands and _ps_trace.tag_key tables have rows managed by the extension and potentially user-supplied rows. For these tables, we manipulate the sequence such that ids < 1000 are managed by the extension and NOT dumped by pg_dump. ids >= 1000 are dumped and restored.

The `_prom_catalog.default` table contains default configuration values. End-users can update these values. The extension inserts the initial defaults during `create extension` and subsequent migrations. We want pg_dump to dump any user-modified values so that they are preserved when restored. We introduced a new view to provide the initial values, and the table's semantics have changed such that they only contain user-modified values. All access to the values now goes through getter/setter functions to handle the merging of the two.

The cascade option is required for dropping the extension. Our new custom type for trace_id is present in the span, link, and event hypertables. This means the custom type is present in the associated compressed tables and chunks. Dropping the extension drops the custom type and the span, link, and event tables, but does not attempt to drop the compressed tables and chunks. Thus, the drop fails because we cannot drop the trace_id type when the compressed tables and chunk definitions depend on it. The cascade option fixes this. This is probably an area for improvement in timescaledb.

The span, link, and event tables are hypertables if timescaledb is installed. The promscale extension creates these tables and various functions and views that depend on these tables. We need the associated backing compressed tables and chunks to be dumped and restored by pg_dump to preserve the data, but we need the tables themselves to be defined as a part of `create extension` so that the dependent objects can be created. Therefore, the span, link, and event tables belong to the extension, but the backing compressed tables and chunks do not. During a restore, the span, link, and event tables are created, but we do not call create_hypertable on them (we can't). We added a promscale_post_restore function which wires up the triggers to span, link, and event tables so that subsequent writes are redirected to the appropriate backing tables.

By default, triggers are added to both the parent table AND child tables. We don't want the child tables (chunk tables) to have the trigger, because that would block the insertion of data. So, in `promscale_post_restore` we add a step to drop the triggers from the chunk tables.
